mdtk/downloaders.py

```diff
@@ -0,0 +1,233 @@
+"""
+Classes to download data from different source. Each class gets data from one
+source. The base data to get is midi. To contribute a new dataset, create a
+new class which extends DataDownloader, and write an accompanying test in
+./tests/test_downloads.py
+"""
+import os
+import sys
+import shutil
+import urllib
+import warnings
+import zipfile
+import glob
+from tqdm import tqdm
+
+
+
+USER_HOME = os.path.expanduser('~')
+DEFAULT_CACHE_PATH = os.path.join(USER_HOME, '.mdtk_cache')
+
+
+
+def download_file(source, dest, verbose=True, overwrite=None):
+        """Get a file from a url and save it locally"""
+        if verbose:
+            print(f'Downloading {source} to {dest}')
+        if os.path.exists(dest):
+            if overwrite is None:
```

Comment:
What is the use case of the overwrite flag being ternary? It adds complication (also to the calling script), and I'm not sure if the difference between False and None is enough to warrant that. I'd prefer to keep it simple (even as far as removing the flag) until we have specific use cases for other functionalities.

Reply:
I was needing this for my testing - I would sometimes need to overwrite cache and other times, when I had already downloaded the big zip, I did not want to.

I agree re complication, but, for now, since I've worked the logic all through, it's kind of effort to remove it!

Default setting of `None` will just warn and not download or overwrite - set to `False` to always overwrite.

Reply:
`True` always overwrites, right? `False` raises an error.

Functionally, it seems `None` assumes the file already there is correct and continues as usual. `False` raises an error and (presumably) kills the calling process (or allows it to handle the error in a custom way).